src/filters/matches.rs

```diff
@@ -0,0 +1,159 @@
+crate::include_proto!("quilkin.extensions.filters.matches.v1alpha1");
+
+use self::quilkin::extensions::filters::matches::v1alpha1::Matches as ProtoConfig;
+use crate::filters::{prelude::*, registry::FilterRegistry};
+
+pub const NAME: &str = "quilkin.extensions.filters.matches.v1alpha1.Matches";
+
+/// Creates a new factory for generating capture filters.
+pub fn factory() -> DynFilterFactory {
+    Box::from(MatchesFactory::new())
+}
+
+struct Matches {
+    config: Config,
+    metrics_registry: prometheus::Registry,
+    filter_registry: FilterRegistry,
+}
+
+impl Matches {
+    fn new(
+        config: Config,
+        filter_registry: FilterRegistry,
+        metrics_registry: prometheus::Registry,
+    ) -> Self {
+        Self {
+            config,
+            filter_registry,
+            metrics_registry,
+        }
+    }
+}
+
+/// The actual behaviour of the matches filter, the only thing that changes
+/// between read and write is which config and direction is called.
+macro_rules! matches_filter {
```

Comment:
Yeah, I can give it a shot rewriting this into a function, it seems like it might be possible, just will require a good chunk of generics.

Reply:
Do we mind if it's mostly copy/paste between `read` and `write`? I know it's duplication, but it's only happening twice, and might just make things more readable? It's not a huge chunk of code.

(Or maybe break things out into several smaller helper functions even?)

Reply:
I think it's important that these have identical behaviour, which is why it shouldn't be duplicated. While it might be easier to ensure now that these two things do the same thing, if we add features to it, it's easy to overtime add it to one side, and not the other. This reduces some of the need for testing because testing one side's behaviour also tests the other side.